apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts

```diff
@@ -0,0 +1,181 @@
+import { IPutFetchOptions, IGetFetchOptions, WebClient } from '../../../utilities/WebClient';
+
+const CONTENT_HASH_HEADER_NAME: 'x-amz-content-sha256' = 'x-amz-content-sha256';
+const DATE_HEADER_NAME: 'x-amz-date' = 'x-amz-date';
+const HOST_HEADER_NAME: 'host' = 'host';
+
+export interface IAmazonS3Credentials {
+  accessKeyId: string;
+  secretAccessKey: string;
+}
+
+interface IIsoDateString {
+  date: string;
+  dateTime: string;
+}
+
+export class AmazonS3Client {
+  private readonly _accessKeyId: string;
+  private readonly _secretAccessKey: string;
+  private readonly _s3Bucket: string;
+  private readonly _s3Region: string;
+
+  private readonly _webClient: WebClient;
+
+  public constructor(
+    credentials: IAmazonS3Credentials | undefined,
+    options: IAmazonS3BuildCacheProviderOptions
+  ) {
+    if (!credentials) {
+      throw new Error('Amazon S3 credential is required.');
+    }
+
+    this._accessKeyId = credentials.accessKeyId || '';
+    this._secretAccessKey = credentials.secretAccessKey || '';
+
+    this._s3Bucket = options.s3Bucket;
+    this._s3Region = options.s3Region;
+
+    this._webClient = new WebClient();
+  }
+
+  public static tryDeserializeCredentials(
+    credentialString: string | undefined
+  ): IAmazonS3Credentials | undefined {
+    if (!credentialString) {
+      return undefined;
+    }
+
+    const splitIndex: number = credentialString.indexOf(':');
+    if (splitIndex === -1) {
+      throw new Error('Amazon S3 credential is in an unexpected format.');
+    }
+
+    return {
+      accessKeyId: credentialString.substring(0, splitIndex),
+      secretAccessKey: credentialString.substring(splitIndex + 1)
+    };
+  }
+
+  public async getObjectAsync(objectName: string): Promise<Buffer | undefined> {
+    const response: fetch.Response = await this._makeRequestAsync('GET', objectName);
+    if (response.ok) {
+      return await response.buffer();
+    } else if (response.status === 404) {
+      return undefined;
+    } else {
+      this._throwS3Error(response);
+    }
+  }
+
+  public async uploadObjectAsync(objectName: string, objectBuffer: Buffer): Promise<void> {
+    const response: fetch.Response = await this._makeRequestAsync('PUT', objectName, objectBuffer);
+    if (!response.ok) {
+      this._throwS3Error(response);
+    }
+  }
+
+  private async _makeRequestAsync(
+    verb: 'GET' | 'PUT',
+    objectName: string,
+    body?: Buffer
+  ): Promise<fetch.Response> {
+    const isoDateString: IIsoDateString = this._getIsoDateString();
+    const bodyHash: string = this._getSha256(body);
+
+    // Compute the authorization header. See https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+    const host: string = `${this._s3Bucket}.s3.amazonaws.com`;
+    const signedHeaderNames: string = `${HOST_HEADER_NAME};${CONTENT_HASH_HEADER_NAME};${DATE_HEADER_NAME}`;
+    const canonicalRequest: string = [
```

Comment:
Could you add a code comment giving an example value for the `canonicalRequest` string that is assembled by this pattern?